Sqlsrv driver: support for limit and offset on MS SQL server.

jakubskrz opened the pull request:
Since MS SQL 2012 there is equivalent to `LIMIT OFFSET` clause in T-SQL.

But it requires to use `ORDER BY` clause in query, see [official docs](https://technet.microsoft.com/en-us/library/gg699618(v=sql.110).aspx)

To consider:
- add default `ORDER BY` instead of exception
- throw `NotSupportedException` when connected to older MSSQL Server version

milo commented:
What about older MSSQL? Sqlsrv driver supports MSSQL since version 2005.

jakubskrz commented:
On 2005 and 2008 versions this SQL syntax cause an error, so it would be good solution to check database version, like this:

```php
$info = $this->connection->query('SELECT SERVERPROPERTY(\'ProductVersion\') AS version')->fetch();

if ($info->version < 11) { // 11 == SQL Server 2012
	throw new Nette\NotSupportedException('Offset is not supported by this database.');
}
```

But that doesn't work for me, I'm getting NULL.

jakubskrz commented:
Hm, OK, when selecting server version via sql query, it's needed to be manually cast to string.

So this one 43a7fc1 will preserve current behaviour on older MS SQL versions.

meridius commented on src/Database/Drivers/SqlsrvDriver.php:
```diff
@@ -91,15 +91,28 @@ public function formatLike($value, $pos)
-			$sql = preg_replace('#^\s*(SELECT|UPDATE|DELETE)#i', '$0 TOP ' . (int) $limit, $sql, 1, $count);
-			if (!$count) {
-				throw new Nette\InvalidArgumentException('SQL query must begin with SELECT, UPDATE or DELETE command.');
+		$info = $this->connection->query("SELECT CAST(SERVERPROPERTY('ProductVersion') AS varchar) AS [ver]")->fetch();
```
How about checking version only once, like this:
```php
	public function applyLimit(& $sql, $limit, $offset)
	{
		static $ver = null;
		if (is_null($ver)) {
			$info = $this->connection->query("SELECT CAST(SERVERPROPERTY('ProductVersion') AS varchar) AS [ver]")->fetch();
			$ver = (int) $info->ver;
		}
		if ($ver < 11) { // 11 == SQL Server 2012
```

I'd really like to see this merged. Perhaps squashing would help?

dg replied:
Is possible to use `$connection->getPdo()->getAttribute(\PDO::ATTR_SERVER_VERSION)` like in MySqlDriver?

meridius replied:
Yes, it is.
```php

	/** @var int */
	private $version;

	public function __construct(Nette\Database\Connection $connection, array $options)
	{
		$this->connection = $connection;
		$this->version = $connection->getPdo()->getAttribute(\PDO::ATTR_SERVER_VERSION);
	}
...
	public function applyLimit(& $sql, $limit, $offset)
	{
		if (version_compare($this->version, 11, '<')) { // < SQL Server 2012
```

dg commented on src/Database/Drivers/SqlsrvDriver.php:
```diff
@@ -91,15 +91,28 @@ public function formatLike($value, $pos)
+				throw new Nette\NotSupportedException('Offset is not supported by this database.');
+			}
+		} else {
+			if (!preg_match('#ORDER\sBY\s(.*)#i', $sql)) {
```
This is not good, what about query `SELECT "ORDER BY"`?

meridius replied:
This kinda makes sense. How about this:
```php
			if (!preg_match('#ORDER\sBY\s(.*)#i', $sql)) {
				throw new Nette\InvalidArgumentException('SQL query must have ORDER BY defined.');
			}
```

dg replied:
I think it is not needed, database will throw exception anyway.

meridius replied:
Yes, but its message is somewhat hard to read: 
`SQL Error (153): Incorrect syntax near '2' Invalid usage of the option next in the FETCH statement. `
for SQL:
`SELECT * FROM address OFFSET 2 ROWS FETCH NEXT 5 ROWS ONLY`

dg replied:
It should be `'#\bORDER\s+BY\b#i'`

meridius commented:
@jakubskrz Could you please incorporate all the mentioned changes and squash all to one commit so it could be merged?

jakubskrz commented:
Hi, sorry for delay, I hadn't much time last week.

I've updated my PR with suggested changes and also with changes from this commit b3b6e9bb73702119a596b4f408bad4356fc53cb1 so there is no merge conflict with current head. Looking at #99, I've also removed checking for `ORDER BY` clause.

But I can't test it right now, as I can't access my dev SQL server (broken VPN :/).

dg commented:
Merged, thx 15ccd8be921dfa65235fb57e574a218a09a9bbcd
